### PR TITLE
Support verifying message metadata via verifier (CLI/FFI)

### PR DESCRIPTION
**Pull request by mefellows**
There is currently no way to pass metadata back to the (HTTP based) verifier, for use by the command line or FFI interface.

This PR adds that.

1. I noticed the verifier doesn't read the `metadata` section (still uses `metaData` when reading a Pact) but generates a pact with `metadata`. It now reads both for compatibility
2. The incoming payload now supports a header (`PACT_MESSAGE_METADATA`) that accepts a base64 encoded JSON of key/value pairs. The values may be complex JSON objects, or simple primitives
3. The `metadata` property of the `Message` struct, now supports complex values to support this.

### CODE REVIEW

**Comment by TimothyJones** [comment]
> I think we could support that being first trying to unmarshal an object of the type above, and falling back to current behaviour otherwise - thoughts?

I think `contents` and `metadata` are too generic, so if we switch behaviour on those names then we'll do the wrong thing sometimes.

Also, I think the example isn't correct: you have

```
{
  "id": 127,
  "event": "some message",
  "__pactMetadata": {
    "Content-Type": "application/json",
    "topic": "baz"
  }
}
```

but to match the implementation in pact-js, it would be:

```
{
  "message": { 
     "id": 127,
     "event": "some message",
  },
  "__pactMessageMetadata": {
    "Content-Type": "application/json",
    "topic": "baz"
  }
}
```

(although now that I read the PR again, perhaps we're not on the same page about the original proposal? I was proposing the wrapped payload, but with less generic field names).

**Comment by TimothyJones** [comment]
This is where I added the wrapper: https://github.com/pact-foundation/pact-js/commit/4aa4e9faf6f35887b365bcb830e2f8ed9eb8c25f#diff-0b4824334fab898c44e8a3dd93c2c7a0408727ef2a1a1e1ca141de5c3c6cd226R37

File: rust/pact_matching/src/models/message.rs
Comment (uglyog) [inline rust/pact_matching/src/models/message.rs:186]: The `get` function returns an optional, so you can use the or functions instead of repeating the code

Comment (mefellows) [reply rust/pact_matching/src/models/message.rs:186]: aha!

File: rust/pact_verifier/src/messages.rs
Comment (TimothyJones) [inline rust/pact_verifier/src/messages.rs:46]: I think `response` might be a bit generic - should we use "internal payload" or something? Otherwise in some cases the logs are going to say "actual response: { ....}" and then "actual message { .... }" which will only be different if you have metadata.

(aside - maybe we should log metadata to `debug` and this whole payload to `trace`, since the message is already logged below?)

File: rust/pact_verifier/src/messages.rs
Comment (uglyog) [inline rust/pact_verifier/src/messages.rs:128]: Just use `contentType` here

File: rust/pact_verifier/src/messages.rs
Comment (TimothyJones) [inline rust/pact_verifier/src/messages.rs:134]: I think this is a `trace` level log.

**Comment by mefellows** [comment]
> I think contents and metadata are too generic, so if we switch behaviour on those names then we'll do the wrong thing sometimes.

They represent the external facing model of a message pact, but I'm open to other suggestions.

**Comment by uglyog** [comment]
I think the idea was to not impact existing users. Now it is just looking for an additional field, but a future well formed payload is better

File: rust/pact_verifier/src/messages.rs
Comment (mefellows) [reply rust/pact_verifier/src/messages.rs:128]: OK. I meant to raise that in the request as well. Can we assume that any consumer adding `Content-Type` as an expectation is satisfied by this?

File: rust/pact_verifier/src/messages.rs
Comment (TimothyJones) [inline rust/pact_verifier/src/messages.rs:50]: See comment directly on the PR - this doesn't match the draft implementation on the client (which does match your proposed improvement ;) ).

File: rust/pact_verifier/src/messages.rs
Comment (uglyog) [reply rust/pact_verifier/src/messages.rs:134]: I think this is a `debug` level log.

**Comment by TimothyJones** [comment]
> They represent the external facing model of a message pact, but I'm open to other suggestions.

I think I'm missing something here. What is the reason that `__pactMessageMetadata` doesn't work?

**Comment by mefellows** [comment]
> I think I'm missing something here. What is the reason that `__pactMessageMetadata` doesn't work?

Because if you return a stream of bytes, or a single JSON primitive as the actual message, you can't add another field to it.

So you need a container to wrap the actual response from the message producer.

Your proposal works, but then I'm not sure why we need the `__` prefix - it's already a Pact specific payload then.

File: rust/pact_verifier/src/messages.rs
Comment (mefellows) [reply rust/pact_verifier/src/messages.rs:50]: yep. I think I'm going to make it accept an object that wraps both the message and the metadata, or fallback to current behaviour. That way we don't break existing clients.

File: rust/pact_verifier/src/messages.rs
Comment (mefellows) [reply rust/pact_verifier/src/messages.rs:46]: I'm going to move this into the `make_provider_request` method alongside the other debugging statements. That way it will be logged for all uses of this important method.

File: rust/pact_verifier/src/messages.rs
Comment (TimothyJones) [reply rust/pact_verifier/src/messages.rs:134]: Following the guidelines we were talking about last week [recorded here](https://github.com/pact-foundation/pact-specification#logging-in-a-pact-library), I think this log is more for maintainers than it is for users (because the whole metadata structure is already logged elsewhere). 

If I'm reading it correctly - which I might not be - this is logging the construction of the metadata structure/object/hash (whatever it's called in rust), which is probably only useful when debugging pact, not when a user is debugging their tests.

Another test is the language in the log message -a user doesn't know what "appending" means in this context, but the maintainer(s) of this code do.

I think the best experience for users would have the metadata logged once, using a log message that requires no extra context.

**Comment by TimothyJones** [comment]
> Because if you return a stream of bytes, or a single JSON primitive as the actual message, you can't add another field to it.
> 
> So you need a container to wrap the actual response from the message producer.

Yes, that's the proposed implementation - I don't think we should be adding extra fields to a user's payload.

> 
> Your proposal works, but then I'm not sure why we need the `__` prefix - it's already a Pact specific payload then.

I suppose `pactMessageMetadata` works just as well, the `__` is only to increase the chance that we're not going to collide with other fields. I think `metadata` doesn't work, because we don't know if we're getting the pact-specific payload or just the raw message content.

File: rust/pact_verifier/src/messages.rs
Comment (mefellows) [reply rust/pact_verifier/src/messages.rs]: Let's leave this for another pass through. I did update another to trace though.

I think related to this, we should read through the info and debug logs for key scenarios and ask ourselves if it presents a coherent and clear view of what's happening. This way, we aren't just changing log levels, but are proactively giving the best experience for users to self diagnose, and we get good feedback if a bug request is raised.

File: rust/pact_matching/src/models/message.rs
Comment (uglyog) [inline rust/pact_matching/src/models/message.rs:301]: This looks strange to me. Is the intent to convert the Value to a string? There is a util function `json_to_string` which does that

File: rust/pact_matching/src/models/message.rs
Comment (uglyog) [inline rust/pact_matching/src/models/message.rs:269]: Can use `json_to_string` function

**Review by uglyog — APPROVED** [review]

File: rust/pact_matching/src/models/message.rs
Comment (mefellows) [reply rust/pact_matching/src/models/message.rs]: Thanks. I'll check to see other uses of that now too.

File: rust/pact_matching/src/lib.rs
Comment (mefellows) [inline rust/pact_matching/src/lib.rs:1426]: @uglyog this look OK to you? (not introduced in this PR but the tests pass and my complex metadata check with Pact Go also passes)

File: rust/pact_matching/src/models/message.rs
Comment (mefellows) [reply rust/pact_matching/src/models/message.rs]: Aha! I knew I'd  seen that pattern elsewhere :P 

Screen Shot 2021-07-06 at 2 31 48 pm

I've found 3 other places that used this same approach before and have changed. Could you please double check?

The tests pass and my complex metadata check with Pact Go also passes, but good to get a second pair of eyes on it.

File: rust/pact_matching/src/lib.rs
Comment (uglyog) [reply rust/pact_matching/src/lib.rs:1426]: LGTM

**Pull request merged**
